Add TLS infos to ClientAuth interface

cdevienne opened the pull request:
It makes it possible to implement a Auth that uses client TLS certificates
to identify them.

derekcollison commented on server/auth.go:
```diff
@@ -12,6 +16,9 @@ type Auth interface {
 	GetOpts() *clientOpts
+	// If TLS is enabled, return true and the TLS ConnectionState
+	// If not, return false
+	GetTLSConnectionState() (bool, tls.ConnectionState)
```
Can you not just return a pointer, which can be nil if TLS is not enabled? Do we need to copy the struct via the return arg?

cdevienne commented:
I updated the PR, which lost the previous commit and its comments altogether :-(

The initial version of GetTLSConnectionState returned (bool, tls.ConnectionState), which @derekcollison suggested to change for a single pointer.

Sorry about the loss of infos, next time I will avoid amending a commit.

derekcollison commented on server/client.go:
```diff
@@ -146,6 +147,15 @@ func (c *client) GetOpts() *clientOpts {
+	if !ok {
+		return nil
+	}
+	state := tc.ConnectionState()
```
Understood, but we do not know that it will be read only. I like the idea of wanting to use TLS as an auth mechanism, just think there may be a better way.

cdevienne replied:
The underlying function (https://golang.org/pkg/crypto/tls/#Conn.ConnectionState) already uses a lock during state preparation.

I think the ConnectionState, in the crypto/tls api, is meant for read-only operations, and modifying the slices in it may lead to serious problems, but nothing is done to avoid that.

Question is: should we protect something crypto/tls does not bother to protect?

cdevienne commented:
I would prefer not to hit the code coverage, but I felt a bit lost in the tests when I had a look.
Any advice on were I can insert something that will cover the new code ?

cdevienne commented:
@derekcollison I added tests that fully cover the new code of my PR.
I don't get why the coverage decreased (on my environment it increased).
Any chance the PR get merged ?

kozlovic commented:
@cdevienne I was asked to look at this PR. I agree that we don't need extra locking since `tls.Conn.ConnectionState()` internally uses its own lock. Furthermore, a copy of the state is returned, so even if the caller was to mess up with the returned state, it would not have an impact on the TLS connection.
The only note I would have is to add comment to `GetTLSConnectionState` since it is an exported method.

cdevienne commented:
Thanks for the review @kozlovic.
Totally agree the function needs a comment (in my own code I do not tolerate non-commented exported functions), I added it and updated the PR.

kozlovic commented:
LGTM
